<a href="https://colab.research.google.com/github/futurebly/Web-AI/blob/main/Day4~10_%ED%8C%90%EB%8B%A4%EC%8A%A4%EC%99%80_%ED%86%B5%EA%B3%84%EA%B8%B0%EC%B4%88.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
########################################## Day 4 ##########################################

# 통계 기초

## 평균

In [ ]:
list(range(10))

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [ ]:
import random
sales = [random.randint(1, 10) for i in range(10)]
sales

[9, 7, 2, 10, 8, 5, 8, 1, 7, 6]

In [ ]:
mean = sum(sales) / len(sales)
mean

6.3

In [ ]:
import pandas as pd
df = pd.DataFrame(sales, columns=['sales'])
df

sales
0      9
1      7
2      2
3     10
4      8
5      5
6      8
7      1
8      7
9      6

In [ ]:
df.mean()

sales    6.3
dtype: float64

## 표준편차

In [ ]:
import random
sales = [random.randint(1, 10) for i in range(10)]
sales

[9, 8, 1, 6, 8, 1, 7, 8, 6, 9]

In [ ]:
mean = sum(sales) / len(sales)
mean

6.3

In [ ]:
deviations = [round(s - mean, 2) for s in sales]
deviations

[2.7, 1.7, -5.3, -0.3, 1.7, -5.3, 0.7, 1.7, -0.3, 2.7]

In [ ]:
import math
분산 = sum([d ** 2 for d in deviations]) / len(deviations)
표준편차 = math.sqrt(분산)
print(분산, 표준편차)

8.010000000000002 2.8301943396169813


In [ ]:
genders = [0,0,0,1,1,1,1,1,1,1]
sum(genders) / len(genders)

0.7

In [ ]:
genders = [1,1,1,0,0,0,0,0,0,0]
sum(genders) / len(genders)

0.3

In [ ]:
import pandas as pd
df = pd.DataFrame(sales, columns=['sales'])
df

sales
0      9
1      8
2      1
3      6
4      8
5      1
6      7
7      8
8      6
9      9

In [ ]:
df['sales'].std()

2.9832867780352594

## 중앙값

In [ ]:
import random
sales = [random.randint(1, 100) for i in range(11)]
sales.sort()
sales

[13, 37, 48, 56, 64, 66, 83, 87, 98, 99, 100]

In [ ]:
# //:몫, %:나머지
n = len(sales) // 2
if len(sales) % 2 == 0:
  print(sum(sales[n-1:n+1]) / 2)
else:
  print(sales[n])

66


In [ ]:
sum(sales) / len(sales)

68.27272727272727

## 사분위수

In [ ]:
import random
sales = [random.randint(1, 100) for i in range(16)]
sales.sort()
sales

[1, 2, 2, 11, 14, 21, 28, 28, 40, 40, 49, 60, 69, 76, 89, 92]

In [ ]:
n = len(sales) // 4
print("count", n)
print(f"{sales[n]}, {sales[2*n]}, {sales[3*n]}")

count 4
14, 40, 69


In [ ]:
import pandas as pd
df = pd.DataFrame(sales, columns=['sales'])
df

sales
0       1
1       2
2       2
3      11
4      14
5      21
6      28
7      28
8      40
9      40
10     49
11     60
12     69
13     76
14     89
15     92

In [ ]:
df['sales'].quantile([0,.25,.5,.75,1])

0.00     1.00
0.25    13.25
0.50    34.00
0.75    62.25
1.00    92.00
Name: sales, dtype: float64

In [ ]:
df.describe()

sales
count  16.000000
mean   38.875000
std    30.852066
min     1.000000
25%    13.250000
50%    34.000000
75%    62.250000
max    92.000000

### 연습문제1. 누구를 기용할 것인가?

In [ ]:
players = {
    'p1': [7, 9, 9, 10, 10, 10, 10, 10, 11, 11, 13],
    'p2': [7, 8, 9, 9, 10, 10, 10, 11, 11, 12, 13],
    'p3': [3, 3, 6, 7, 7, 10, 10, 10, 11, 13, 30]
}
import pandas as pd
df = pd.DataFrame(players)
df

p1  p2  p3
0    7   7   3
1    9   8   3
2    9   9   6
3   10   9   7
4   10  10   7
5   10  10  10
6   10  10  10
7   10  11  10
8   11  11  11
9   11  12  13
10  13  13  30

In [ ]:
df.describe()

p1         p2         p3
count  11.00000  11.000000  11.000000
mean   10.00000  10.000000  10.000000
std     1.48324   1.732051   7.362065
min     7.00000   7.000000   3.000000
25%     9.50000   9.000000   6.500000
50%    10.00000  10.000000  10.000000
75%    10.50000  11.000000  10.500000
max    13.00000  13.000000  30.000000

## list comprehension

In [ ]:
numbers = list(range(10))
numbers

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [ ]:
evens = [n for n in numbers if n % 2 == 0]
evens

[0, 2, 4, 6, 8]

In [ ]:
evens = []
for n in numbers:
  if n % 2 == 0:
    evens.append(n)
evens

[0, 2, 4, 6, 8]

In [ ]:
odds = [n for n in numbers if n % 2 != 0]
odds

[1, 3, 5, 7, 9]

In [ ]:
squre_evens = [n ** 2 for n in numbers if n % 2 == 0]
squre_evens

[0, 4, 16, 36, 64]

## slicing - 부분집합 꺼내기

In [ ]:
numbers = list(range(10))
numbers

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [ ]:
numbers[3:6]

[3, 4, 5]

In [ ]:
numbers[:6]

[0, 1, 2, 3, 4, 5]

In [ ]:
numbers[6:]

[6, 7, 8, 9]

In [ ]:
numbers[-3:]

[7, 8, 9]

In [ ]:
########################################## Day 5 ##########################################

## 최빈값

In [ ]:
fruit = ['배', '배', '배', '배', '사과', '사과', '귤', '귤', '귤']
set(fruit)

{'귤', '배', '사과'}

In [ ]:
fruit_count = [(s, fruit.count(s)) for s in set(fruit)]
fruit_count

[('배', 4), ('귤', 3), ('사과', 2)]

In [ ]:
fruit_count = dict([(s, fruit.count(s)) for s in set(fruit)])
fruit_count

{'귤': 3, '배': 4, '사과': 2}

In [ ]:
import pandas as pd
df = pd.DataFrame(fruit, columns=['fruit'])
df

fruit
0     배
1     배
2     배
3     배
4    사과
5    사과
6     귤
7     귤
8     귤

In [ ]:
df['fruit'].value_counts()

배     4
귤     3
사과    2
Name: fruit, dtype: int64

In [ ]:
df['fruit'].value_counts(normalize=True) #그 확률

배     0.444444
귤     0.333333
사과    0.222222
Name: fruit, dtype: float64

# 판다스 기초

## 데이터 불러오기/파악하기

- 레몬에이드 데이터

In [ ]:
import pandas as pd
path = 'https://raw.githubusercontent.com/blackdew/ml-tensorflow/master/data/csv/lemonade.csv'
df = pd.read_csv(path)
print(df.shape)
df.info()

(6, 2)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6 entries, 0 to 5
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   온도      6 non-null      int64
 1   판매량     6 non-null      int64
dtypes: int64(2)
memory usage: 224.0 bytes


In [ ]:
df.head(2)

온도  판매량
0  20   40
1  21   42

In [ ]:
df.tail(2)

온도  판매량
4  24   48
5  25   50

In [ ]:
df.columns

Index(['온도', '판매량'], dtype='object')

- 보스턴 집값 데이터

In [ ]:
import pandas as pd
path = 'https://raw.githubusercontent.com/blackdew/ml-tensorflow/master/data/csv/boston.csv'
df = pd.read_csv(path)
print(df.shape)
df.info()

(506, 14)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   crim     506 non-null    float64
 1   zn       506 non-null    float64
 2   indus    506 non-null    float64
 3   chas     506 non-null    int64  
 4   nox      506 non-null    float64
 5   rm       506 non-null    float64
 6   age      506 non-null    float64
 7   dis      506 non-null    float64
 8   rad      506 non-null    int64  
 9   tax      506 non-null    int64  
 10  ptratio  506 non-null    float64
 11  b        506 non-null    float64
 12  lstat    506 non-null    float64
 13  medv     506 non-null    float64
dtypes: float64(11), int64(3)
memory usage: 55.5 KB


In [ ]:
df.head(2)

crim    zn  indus  chas    nox  ...  tax  ptratio      b  lstat  medv
0  0.00632  18.0   2.31     0  0.538  ...  296     15.3  396.9   4.98  24.0
1  0.02731   0.0   7.07     0  0.469  ...  242     17.8  396.9   9.14  21.6

[2 rows x 14 columns]

In [ ]:
df.tail(2)

crim   zn  indus  chas    nox  ...  tax  ptratio       b  lstat  medv
504  0.10959  0.0  11.93     0  0.573  ...  273     21.0  393.45   6.48  22.0
505  0.04741  0.0  11.93     0  0.573  ...  273     21.0  396.90   7.88  11.9

[2 rows x 14 columns]

In [ ]:
df.columns

Index(['crim', 'zn', 'indus', 'chas', 'nox', 'rm', 'age', 'dis', 'rad', 'tax',
       'ptratio', 'b', 'lstat', 'medv'],
      dtype='object')

- 아이리스 품종 데이터

In [ ]:
import pandas as pd
path = 'https://raw.githubusercontent.com/blackdew/ml-tensorflow/master/data/csv/iris.csv'
df = pd.read_csv(path)
print(df.shape)
df.info()

(150, 5)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   꽃잎길이    150 non-null    float64
 1   꽃잎폭     150 non-null    float64
 2   꽃받침길이   150 non-null    float64
 3   꽃받침폭    150 non-null    float64
 4   품종      150 non-null    object 
dtypes: float64(4), object(1)
memory usage: 6.0+ KB


In [ ]:
df.head(2)

꽃잎길이  꽃잎폭  꽃받침길이  꽃받침폭      품종
0   5.1  3.5    1.4   0.2  setosa
1   4.9  3.0    1.4   0.2  setosa

In [ ]:
df.tail(2)

꽃잎길이  꽃잎폭  꽃받침길이  꽃받침폭         품종
148   6.2  3.4    5.4   2.3  virginica
149   5.9  3.0    5.1   1.8  virginica

In [ ]:
df.columns

Index(['꽃잎길이', '꽃잎폭', '꽃받침길이', '꽃받침폭', '품종'], dtype='object')

## Row/Column 선택하기

In [ ]:
import pandas as pd
path = 'https://raw.githubusercontent.com/blackdew/ml-tensorflow/master/data/csv/lemonade.csv'
df = pd.read_csv(path)
print(df['온도'].shape)
df['온도']

(6,)


0    20
1    21
2    22
3    23
4    24
5    25
Name: 온도, dtype: int64

In [ ]:
print(df[['온도']].shape)
df[['온도']]

(6, 1)


온도
0  20
1  21
2  22
3  23
4  24
5  25

In [ ]:
df[['온도', '판매량']]

온도  판매량
0  20   40
1  21   42
2  22   44
3  23   46
4  24   48
5  25   50

In [ ]:
df.loc[:3]

온도  판매량
0  20   40
1  21   42
2  22   44
3  23   46

In [ ]:
df.iloc[:3]

온도  판매량
0  20   40
1  21   42
2  22   44

In [ ]:
df.loc[2:4]

온도  판매량
2  22   44
3  23   46
4  24   48

In [ ]:
df.iloc[2:4]

온도  판매량
2  22   44
3  23   46

In [ ]:
df.iloc[3,1]

46

In [ ]:
df.iloc[2:4, :]

온도  판매량
2  22   44
3  23   46

In [ ]:
df.iloc[2:4, 1:]

판매량
2   44
3   46

In [ ]:
df_sorted = df.sort_values(by=['판매량'], ascending=False)
df_sorted.head()

온도  판매량
5  25   50
4  24   48
3  23   46
2  22   44
1  21   42

In [ ]:
df_sorted.iloc[0]

온도     25
판매량    50
Name: 5, dtype: int64

In [ ]:
df_sorted.loc[0]

온도     20
판매량    40
Name: 0, dtype: int64

## 원하는 데이터 찾기 - Query/Search

In [ ]:
import pandas as pd
path = 'https://raw.githubusercontent.com/blackdew/ml-tensorflow/master/data/csv/lemonade.csv'
df = pd.read_csv(path)
df

온도  판매량
0  20   40
1  21   42
2  22   44
3  23   46
4  24   48
5  25   50

In [ ]:
df['온도'] > 23

0    False
1    False
2    False
3    False
4     True
5     True
Name: 온도, dtype: bool

In [ ]:
df[df['온도'] > 23]

온도  판매량
4  24   48
5  25   50

In [ ]:
df[(df['온도'] > 23) | (df['판매량'] == 40)] # |: or

온도  판매량
0  20   40
4  24   48
5  25   50

In [ ]:
d = {'col1': [0, 1, 2, 3],
     'col2': ['Gold', 'Bronze', 'Gold', 'Silver']}
df1 = pd.DataFrame(d)
df1

col1    col2
0     0    Gold
1     1  Bronze
2     2    Gold
3     3  Silver

In [ ]:
df1['col2'].str.contains('o')

0     True
1     True
2     True
3    False
Name: col2, dtype: bool

In [ ]:
df1[df1['col2'].str.contains('o')]

col1    col2
0     0    Gold
1     1  Bronze
2     2    Gold

### 연습문제1. chas가 1인 데이터 중에 가장 비싼 집을 찾아봅시다.

In [ ]:
import pandas as pd
path = 'https://raw.githubusercontent.com/blackdew/ml-tensorflow/master/data/csv/boston.csv'
df = pd.read_csv(path)
df.head()

crim    zn  indus  chas    nox  ...  tax  ptratio       b  lstat  medv
0  0.00632  18.0   2.31     0  0.538  ...  296     15.3  396.90   4.98  24.0
1  0.02731   0.0   7.07     0  0.469  ...  242     17.8  396.90   9.14  21.6
2  0.02729   0.0   7.07     0  0.469  ...  242     17.8  392.83   4.03  34.7
3  0.03237   0.0   2.18     0  0.458  ...  222     18.7  394.63   2.94  33.4
4  0.06905   0.0   2.18     0  0.458  ...  222     18.7  396.90   5.33  36.2

[5 rows x 14 columns]

In [ ]:
df_selected = df[(df['chas'] == 1)]
df_selected.head()

crim   zn  indus  chas    nox  ...  tax  ptratio       b  lstat  medv
142  3.32105  0.0  19.58     1  0.871  ...  403     14.7  396.90  26.82  13.4
152  1.12658  0.0  19.58     1  0.871  ...  403     14.7  343.28  12.12  15.3
154  1.41385  0.0  19.58     1  0.871  ...  403     14.7  321.02  15.12  17.0
155  3.53501  0.0  19.58     1  0.871  ...  403     14.7   88.01  15.02  15.6
160  1.27346  0.0  19.58     1  0.605  ...  403     14.7  338.92   5.50  27.0

[5 rows x 14 columns]

In [ ]:
df_selected.sort_values(by=['medv'], ascending=False).head()

crim    zn  indus  chas    nox  ...  tax  ptratio       b  lstat  medv
372  8.26725   0.0  18.10     1  0.668  ...  666     20.2  347.88   8.88  50.0
370  6.53876   0.0  18.10     1  0.631  ...  666     20.2  392.05   2.96  50.0
369  5.66998   0.0  18.10     1  0.631  ...  666     20.2  375.33   3.73  50.0
283  0.01501  90.0   1.21     1  0.401  ...  198     13.6  395.52   3.16  50.0
162  1.83377   0.0  19.58     1  0.605  ...  403     14.7  389.61   1.92  50.0

[5 rows x 14 columns]

### 연습문제2. 품종별로 꽃잎폭이 가장 큰 행을 찾아 비교해 봅시다.

In [ ]:
import pandas as pd
path = 'https://raw.githubusercontent.com/blackdew/ml-tensorflow/master/data/csv/iris.csv'
df = pd.read_csv(path)
df.head()

꽃잎길이  꽃잎폭  꽃받침길이  꽃받침폭      품종
0   5.1  3.5    1.4   0.2  setosa
1   4.9  3.0    1.4   0.2  setosa
2   4.7  3.2    1.3   0.2  setosa
3   4.6  3.1    1.5   0.2  setosa
4   5.0  3.6    1.4   0.2  setosa

In [ ]:
df['품종'].value_counts()

virginica     50
versicolor    50
setosa        50
Name: 품종, dtype: int64

In [ ]:
df_selected = df[(df['품종'] == 'setosa')]
df_selected.sort_values(by=['꽃잎폭'], ascending=False).head(1)

꽃잎길이  꽃잎폭  꽃받침길이  꽃받침폭      품종
15   5.7  4.4    1.5   0.4  setosa

In [ ]:
df_selected = df[(df['품종'] == 'virginica')]
df_selected.sort_values(by=['꽃잎폭'], ascending=False).head(1)

꽃잎길이  꽃잎폭  꽃받침길이  꽃받침폭         품종
131   7.9  3.8    6.4   2.0  virginica

In [ ]:
df_selected = df[(df['품종'] == 'versicolor')]
df_selected.sort_values(by=['꽃잎폭'], ascending=False).head(1)

꽃잎길이  꽃잎폭  꽃받침길이  꽃받침폭          품종
85   6.0  3.4    4.5   1.6  versicolor

In [ ]:
import pandas as pd
path = "/content/sample_data/california_housing_test.csv"
df = pd.read_csv(path)
print(df.shape)
df.head()

(3000, 9)


longitude  latitude  ...  median_income  median_house_value
0    -122.05     37.37  ...         6.6085            344700.0
1    -118.30     34.26  ...         3.5990            176500.0
2    -117.81     33.78  ...         5.7934            270500.0
3    -118.36     33.82  ...         6.1359            330000.0
4    -119.67     36.33  ...         2.9375             81700.0

[5 rows x 9 columns]

## KRX 증권 데이터

In [ ]:
# 라이브러리를 설치할 때에는 맨앞에 !를 붙여야 실행이 됩니다. 
!pip install finance-datareader

In [ ]:
import FinanceDataReader as fdr

# 삼성전자(005930) 전체 (1996-11-05 ~ 현재)
df = fdr.DataReader('005930')
print(df.shape)
df.info()

(6000, 6)
<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 6000 entries, 1997-10-16 to 2021-11-12
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Open    6000 non-null   int64  
 1   High    6000 non-null   int64  
 2   Low     6000 non-null   int64  
 3   Close   6000 non-null   int64  
 4   Volume  6000 non-null   int64  
 5   Change  5999 non-null   float64
dtypes: float64(1), int64(5)
memory usage: 328.1 KB


In [ ]:
df.tail()

Open   High    Low  Close    Volume    Change
Date                                                      
2021-11-08  70400  70600  69800  70600  11121981  0.005698
2021-11-09  70300  71000  70100  70500  11159589 -0.001416
2021-11-10  70300  70900  70200  70200   9422009 -0.004255
2021-11-11  69900  70000  69600  69900  11489122 -0.004274
2021-11-12  70200  70900  69900  70600  10038131  0.010014

In [ ]:
import FinanceDataReader as fdr

# Apple(AAPL), 2017-01-01 ~ Now
df = fdr.DataReader('AAPL', '2017')
df.tail()

Close    Open    High     Low      Volume  Change
Date                                                          
2021-11-08  150.44  151.41  151.57  150.16  54770000.0 -0.0056
2021-11-09  150.81  150.20  151.43  150.06  56570000.0  0.0025
2021-11-10  147.92  150.02  150.13  147.85  64350000.0 -0.0192
2021-11-11  147.87  148.96  149.43  147.68  40720000.0 -0.0003
2021-11-12  149.99  148.43  150.40  147.48  63800000.0  0.0143

In [ ]:
########################################## Day 6 ##########################################

## 그루핑 / 집계하기 (.groupby())

In [ ]:
import pandas as pd
path = 'https://raw.githubusercontent.com/blackdew/ml-tensorflow/master/data/csv/iris.csv'
df = pd.read_csv(path)
df.head()

꽃잎길이  꽃잎폭  꽃받침길이  꽃받침폭      품종
0   5.1  3.5    1.4   0.2  setosa
1   4.9  3.0    1.4   0.2  setosa
2   4.7  3.2    1.3   0.2  setosa
3   4.6  3.1    1.5   0.2  setosa
4   5.0  3.6    1.4   0.2  setosa

In [ ]:
df.describe() #집계하기

꽃잎길이         꽃잎폭       꽃받침길이        꽃받침폭
count  150.000000  150.000000  150.000000  150.000000
mean     5.843333    3.054000    3.758667    1.198667
std      0.828066    0.433594    1.764420    0.763161
min      4.300000    2.000000    1.000000    0.100000
25%      5.100000    2.800000    1.600000    0.300000
50%      5.800000    3.000000    4.350000    1.300000
75%      6.400000    3.300000    5.100000    1.800000
max      7.900000    4.400000    6.900000    2.500000

In [ ]:
print("sum()")
print(df[['꽃잎길이', '꽃잎폭']].sum())

print("mean()")
print(df[['꽃잎길이', '꽃잎폭']].mean())

print("std()")
print(df[['꽃잎길이', '꽃잎폭']].std())

print("median()")
print(df[['꽃잎길이', '꽃잎폭']].median())

print("min()")
print(df[['꽃잎길이', '꽃잎폭']].min())

print("max()")
print(df[['꽃잎길이', '꽃잎폭']].max())

sum()
꽃잎길이    876.5
꽃잎폭     458.1
dtype: float64
mean()
꽃잎길이    5.843333
꽃잎폭     3.054000
dtype: float64
std()
꽃잎길이    0.828066
꽃잎폭     0.433594
dtype: float64
median()
꽃잎길이    5.8
꽃잎폭     3.0
dtype: float64
min()
꽃잎길이    4.3
꽃잎폭     2.0
dtype: float64
max()
꽃잎길이    7.9
꽃잎폭     4.4
dtype: float64


In [ ]:
df.groupby(['품종'])[['꽃잎길이', '꽃잎폭']].mean()

꽃잎길이    꽃잎폭
품종                      
setosa      5.006  3.418
versicolor  5.936  2.770
virginica   6.588  2.974

In [ ]:
df.groupby(['품종'])[['꽃잎길이', '꽃잎폭']].std()

꽃잎길이       꽃잎폭
품종                            
setosa      0.352490  0.381024
versicolor  0.516171  0.313798
virginica   0.635880  0.322497

In [ ]:
df.groupby(['품종'])[['꽃잎길이', '꽃잎폭']].max()

꽃잎길이  꽃잎폭
품종                   
setosa       5.8  4.4
versicolor   7.0  3.4
virginica    7.9  3.8

In [ ]:
df.groupby(['품종'])[['꽃잎길이', '꽃잎폭']].describe()

꽃잎길이                               ...  꽃잎폭                        
           count   mean       std  min    25%  ...  min    25%  50%    75%  max
품종                                             ...                             
setosa      50.0  5.006  0.352490  4.3  4.800  ...  2.3  3.125  3.4  3.675  4.4
versicolor  50.0  5.936  0.516171  4.9  5.600  ...  2.0  2.525  2.8  3.000  3.4
virginica   50.0  6.588  0.635880  4.9  6.225  ...  2.2  2.800  3.0  3.175  3.8

[3 rows x 16 columns]

### 연습문제 1.chas가 1인 데이터들과 0인 데이터의 집값을 집계

In [ ]:
import pandas as pd
path = 'https://raw.githubusercontent.com/blackdew/ml-tensorflow/master/data/csv/boston.csv'
df = pd.read_csv(path)
df.head()

crim    zn  indus  chas    nox  ...  tax  ptratio       b  lstat  medv
0  0.00632  18.0   2.31     0  0.538  ...  296     15.3  396.90   4.98  24.0
1  0.02731   0.0   7.07     0  0.469  ...  242     17.8  396.90   9.14  21.6
2  0.02729   0.0   7.07     0  0.469  ...  242     17.8  392.83   4.03  34.7
3  0.03237   0.0   2.18     0  0.458  ...  222     18.7  394.63   2.94  33.4
4  0.06905   0.0   2.18     0  0.458  ...  222     18.7  396.90   5.33  36.2

[5 rows x 14 columns]

In [ ]:
df.groupby(['chas'])[['medv']].describe()

medv                                                     
      count       mean        std   min   25%   50%    75%   max
chas                                                            
0     471.0  22.093843   8.831362   5.0  16.6  20.9  24.80  50.0
1      35.0  28.440000  11.816643  13.4  21.1  23.3  33.15  50.0

### 연습문제 2. 품종별로 꽃잎, 꽃받침을 집계

In [ ]:
import pandas as pd
path = 'https://raw.githubusercontent.com/blackdew/ml-tensorflow/master/data/csv/iris.csv'
df = pd.read_csv(path)
df.head()

꽃잎길이  꽃잎폭  꽃받침길이  꽃받침폭      품종
0   5.1  3.5    1.4   0.2  setosa
1   4.9  3.0    1.4   0.2  setosa
2   4.7  3.2    1.3   0.2  setosa
3   4.6  3.1    1.5   0.2  setosa
4   5.0  3.6    1.4   0.2  setosa

In [ ]:
df.groupby(['품종'])[['꽃잎길이', '꽃잎폭']].describe()

꽃잎길이                               ...  꽃잎폭                        
           count   mean       std  min    25%  ...  min    25%  50%    75%  max
품종                                             ...                             
setosa      50.0  5.006  0.352490  4.3  4.800  ...  2.3  3.125  3.4  3.675  4.4
versicolor  50.0  5.936  0.516171  4.9  5.600  ...  2.0  2.525  2.8  3.000  3.4
virginica   50.0  6.588  0.635880  4.9  6.225  ...  2.2  2.800  3.0  3.175  3.8

[3 rows x 16 columns]

In [ ]:
df.groupby(['품종'])[['꽃받침길이', '꽃받침폭']].describe()

꽃받침길이                             ... 꽃받침폭                    
           count   mean       std  min  25%  ...  min  25%  50%  75%  max
품종                                           ...                         
setosa      50.0  1.464  0.173511  1.0  1.4  ...  0.1  0.2  0.2  0.3  0.6
versicolor  50.0  4.260  0.469911  3.0  4.0  ...  1.0  1.2  1.3  1.5  1.8
virginica   50.0  5.552  0.551895  4.5  5.1  ...  1.4  1.8  2.0  2.3  2.5

[3 rows x 16 columns]

### 연습문제3. 쿼리를 통해 데이터를 선택하여 집계

In [ ]:
import pandas as pd
path = 'https://raw.githubusercontent.com/blackdew/ml-tensorflow/master/data/csv/boston.csv'
df = pd.read_csv(path)
df.head()

crim    zn  indus  chas    nox  ...  tax  ptratio       b  lstat  medv
0  0.00632  18.0   2.31     0  0.538  ...  296     15.3  396.90   4.98  24.0
1  0.02731   0.0   7.07     0  0.469  ...  242     17.8  396.90   9.14  21.6
2  0.02729   0.0   7.07     0  0.469  ...  242     17.8  392.83   4.03  34.7
3  0.03237   0.0   2.18     0  0.458  ...  222     18.7  394.63   2.94  33.4
4  0.06905   0.0   2.18     0  0.458  ...  222     18.7  396.90   5.33  36.2

[5 rows x 14 columns]

In [ ]:
df['age'].mean()

68.57490118577078

In [ ]:
df[(df['age'] < 68.5)]['medv'].describe()

count    213.000000
mean      26.219249
std        7.429686
min       10.200000
25%       21.200000
50%       23.900000
75%       29.600000
max       50.000000
Name: medv, dtype: float64

In [ ]:
df[(df['age'] > 68.5)]['medv'].describe()

count    293.000000
mean      19.852901
std        9.438864
min        5.000000
25%       13.900000
50%       18.300000
75%       22.600000
max       50.000000
Name: medv, dtype: float64

In [ ]:
df['crim'].mean()

3.6135235573122535

In [ ]:
df[(df['crim'] < 3.6)]['medv'].describe()

count    378.000000
mean      24.677513
std        8.372866
min        7.000000
25%       19.400000
50%       22.600000
75%       28.350000
max       50.000000
Name: medv, dtype: float64

In [ ]:
df[(df['crim'] > 3.6)]['medv'].describe()

count    128.000000
mean      16.199219
std        8.609375
min        5.000000
25%       10.975000
50%       14.150000
75%       19.200000
max       50.000000
Name: medv, dtype: float64

In [ ]:
df[(df['crim'] > 3.6)]['age'].describe()

count    128.000000
mean      91.140625
std       10.504955
min       51.900000
25%       87.900000
50%       94.750000
75%       98.950000
max      100.000000
Name: age, dtype: float64

In [ ]:
df[(df['crim'] < 3.6)]['age'].describe()

count    378.000000
mean      60.933598
std       28.156236
min        2.900000
25%       36.600000
50%       62.650000
75%       88.150000
max      100.000000
Name: age, dtype: float64

### 연습문제 4. 알게된 정보와 떠오르는 질문을 기록해두자.
- age의 평균은 68.4
- age가 68.4 보다 큰 경우 집값 평균은 19.85
- age가 68.4 보다 작은 경우 집값 평균은 26.21

* 범죄율 평균은 3.6
* 범죄율이 평균보다 높은 곳의 집값 평균은 24.67
* 반대의 경우는 16.19
* age가 높을수록 범죄율이 높다
* 범죄율 평균 이상 평균 나이는 91세
* 범죄율 평균 이하 평균 나이는 60세

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   crim     506 non-null    float64
 1   zn       506 non-null    float64
 2   indus    506 non-null    float64
 3   chas     506 non-null    int64  
 4   nox      506 non-null    float64
 5   rm       506 non-null    float64
 6   age      506 non-null    float64
 7   dis      506 non-null    float64
 8   rad      506 non-null    int64  
 9   tax      506 non-null    int64  
 10  ptratio  506 non-null    float64
 11  b        506 non-null    float64
 12  lstat    506 non-null    float64
 13  medv     506 non-null    float64
dtypes: float64(11), int64(3)
memory usage: 55.5 KB


### anscombe data (column 이름 바꾸기)

In [ ]:
import pandas as pd
df = pd.read_json("/content/sample_data/anscombe.json")
df.head()

Series   X     Y
0      I  10  8.04
1      I   8  6.95
2      I  13  7.58
3      I   9  8.81
4      I  11  8.33

In [ ]:
# 칼럼명 바꾸기
df.columns = ['season', 'ad', 'sales']
df.head()

season  ad  sales
0      I  10   8.04
1      I   8   6.95
2      I  13   7.58
3      I   9   8.81
4      I  11   8.33

## 합치기 / _concat

In [ ]:
import pandas as pd
df1 = pd.DataFrame([[1, 2, 3, 4], [4, 5, 6, 7], [7, 8, 9, 10]])
df2 = pd.DataFrame([[1, 2, 3, 4], [4, 5, 6, 7], [7, 8, 9, 10]])
print(df1.shape, df2.shape)

(3, 4) (3, 4)


In [ ]:
df2

0  1  2   3
0  1  2  3   4
1  4  5  6   7
2  7  8  9  10

In [ ]:
 #shape 기준으로 (,)에서 0번째인 행 숫자를 바꾸라는 말
df_concat = pd.concat([df1, df2], axis=0) 
print(df_concat.shape)
df_concat

(6, 4)


0  1  2   3
0  1  2  3   4
1  4  5  6   7
2  7  8  9  10
0  1  2  3   4
1  4  5  6   7
2  7  8  9  10

In [ ]:
df_concat = df_concat.reset_index(drop=True)
df_concat

0  1  2   3
0  1  2  3   4
1  4  5  6   7
2  7  8  9  10
3  1  2  3   4
4  4  5  6   7
5  7  8  9  10

In [ ]:
#axis:축/axis=1은 1번째 축(열)을 바꾸라는 말
df_concat = pd.concat([df1, df2], axis=1)
print(df_concat.shape)
df_concat

(3, 8)


0  1  2   3  0  1  2   3
0  1  2  3   4  1  2  3   4
1  4  5  6   7  4  5  6   7
2  7  8  9  10  7  8  9  10

In [ ]:
df_concat.columns = [0, 1, 2, 3, 4, 5, 6, 7]
df_concat

0  1  2   3  4  5  6   7
0  1  2  3   4  1  2  3   4
1  4  5  6   7  4  5  6   7
2  7  8  9  10  7  8  9  10

In [ ]:
print(df_concat.mean(axis=0))
df_concat.mean(axis=0).shape

0    4.0
1    5.0
2    6.0
3    7.0
4    4.0
5    5.0
6    6.0
7    7.0
dtype: float64


(8,)

In [ ]:
print(df_concat.mean(axis=1))
print(df_concat.mean(axis=1).shape)

0    2.5
1    5.5
2    8.5
dtype: float64
(3,)


In [ ]:
########################################## Day 10 ##########################################

# 비교와 확률

## 검정

### 이항 검정

In [ ]:
from scipy import stats

print('stats.binom_test(8, 10) >', stats.binom_test(8, 10))
print('stats.binom_test(9, 10) >', stats.binom_test(9, 10))
print('stats.binom_test(14, 20) >', stats.binom_test(14, 20))
print('stats.binom_test(15, 20) >', stats.binom_test(15, 20))
print('stats.binom_test(60, 100) >', stats.binom_test(60, 100))
print('stats.binom_test(65, 100) >', stats.binom_test(65, 100))

stats.binom_test(8, 10) > 0.10937500000000003
stats.binom_test(9, 10) > 0.021484374999999997
stats.binom_test(14, 20) > 0.11531829833984371
stats.binom_test(15, 20) > 0.04138946533203125
stats.binom_test(60, 100) > 0.05688793364098078
stats.binom_test(65, 100) > 0.0035176417229701613


### T 검정

In [ ]:
from scipy import stats

#평균이 50이고 표준편차가 2인 집단에서 랜덤하게 20개의 값을 꺼내라.
x1 = stats.norm(50, 2).rvs(20).round(2)
x2 = stats.norm(49, 2).rvs(20).round(2)
print(stats.ttest_ind(x1, x2))
print('평균 : ', round(x1.mean(), 2), round(x2.mean(), 2))
print('표준편차 : ', round(x1.std(), 2), round(x2.std(), 2))
print(x1)
print(x2)
#pvalue는 두 집단이 같은 모집단에서 뽑혔을 확률.
#pvalue가 0.05보다 작으면 x1, x2는 서로 다른 모집단일것.

Ttest_indResult(statistic=0.34202730872674436, pvalue=0.7342148085473583)
평균 :  49.27 49.01
표준편차 :  2.1 2.5
[50.01 51.83 51.54 47.82 48.99 46.09 49.95 52.11 51.61 48.95 48.61 45.69
 46.48 47.   49.44 51.89 50.6  51.52 46.21 49.06]
[48.36 48.38 45.7  49.07 48.43 48.9  48.32 45.79 47.5  44.75 51.37 49.93
 49.47 48.59 48.97 49.46 57.16 51.25 50.6  48.28]


In [ ]:
from scipy import stats

x1 = stats.norm(450, 50).rvs(30).round(0)
x2 = stats.norm(400, 50).rvs(30).round(0)
print(stats.ttest_ind(x1, x2))
print('평균 : ', round(x1.mean(), 2), round(x2.mean(), 2))
print('표준편차 : ', round(x1.std(), 2), round(x2.std(), 2))
print(x1)
print(x2)

Ttest_indResult(statistic=3.0867485798453176, pvalue=0.003101085608390514)
평균 :  437.8 393.5
표준편차 :  40.22 66.0
[405. 419. 518. 460. 438. 465. 386. 386. 406. 518. 380. 415. 407. 454.
 470. 407. 446. 424. 474. 487. 480. 480. 415. 444. 433. 469. 420. 406.
 473. 349.]
[365. 304. 496. 301. 421. 396. 528. 423. 378. 314. 526. 453. 471. 346.
 489. 415. 384. 452. 386. 360. 338. 347. 320. 373. 378. 286. 345. 458.
 338. 414.]
